Expose grammar of graphics symbology to the Python API (#1420)
geojupyter/jupytergis: docs/user_guide/python_api.md

# Python API

## GISDocument

The `GISDocument` class is the main entry point for map creation and editing.

> ### Create a new empty document with map view options

Reviewer: ```suggestion
### Create a new ephemeral document with map view options

If you don't pass a filename as the first positional argument, the document will not be saved to disk.
```

```python
from jupytergis import GISDocument

doc = GISDocument(
    latitude=47.0,
    longitude=2.0,
    zoom=5,
    projection="EPSG:3857",
)

doc
```

> 

Reviewer: ```suggestion

If this document doesn't already exist, it will be created on disk.

```

### Open an existing document

It is possible to open an existing document and start editing it from the Python notebook.

The Python runtime then acts as a collaborator on the document (in the live collaboration sense), so other people can see the modifications being made to the document.

```python
from jupytergis import GISDocument

doc = GISDocument("mydocument.jGIS")

doc

> ```

Reviewer: ````suggestion
```

If this document doesn't exist, it will be created on disk.
````

### Add base layers and vector/raster data

```python
from jupytergis import GISDocument, stroke

doc = GISDocument()

doc.add_raster_layer(
    url="https://tile.openstreetmap.org/{z}/{x}/{y}.png"
)
doc.add_geojson_layer(
    name="Ship Trajectories",
    path="https://openlayers.org/en/latest/examples/data/geojson/ship-trajectories.json",
    symbology=stroke("red")
)
doc.add_geotiff_layer(
    name="Sentinel-2 cloudless",
    url="https://s2downloads.eox.at/demo/EOxCloudless/2020/rgbnir/s2cloudless2020-16bits_sinlge-file_z0-4.tif",
    min=2000,
    max=25000,
)

doc
```

### Add GeoJSON from in-memory data

```python
from jupytergis import GISDocument, stroke, constant_num

latitude = 48.853757
longitude = 2.358213

doc = GISDocument(zoom=12, latitude=latitude, longitude=longitude)

geojson = {
    "type": "FeatureCollection",
    "features": [
        {
            "type": "Feature",
            "geometry": {"type": "Point", "coordinates": [longitude, latitude]},
            "properties": {"name": "Paris", "mag": 4.2},
        },
    ],
}

doc.add_geojson_layer(
    name="Paris",
    data=geojson,
    symbology=stroke("blue", stroke_width=3).constant_num(20, target="radius")

> )

Reviewer: The design you're presenting is wonderful _and_ it's inspiring more ideas. I'd love to hear what you think. I feel bad that I'm doing a sort of "swoop and poop" here at the late stage of this work. I'm sorry 😬

A point of confusion for me: I've been thinking of this method chaining as always having a 1:1 mapping to the "symbology mappings" in the UI. So I initially expected this particular example to look something like:

```python
symbology=[
    constant_num(20, target="radius"),
    constant_color("blue", target="stroke-color"),
    constant_num(3, target="stroke-width"),
],
```

But I think I had a poor mental model. What you had demonstrated in our last meeting was each element in the symbology list corresponding to a "symbology layer" (e.g. a heatmap layer and a regular stroke/fill layer).

Even after rectifying my mental model, I'm still a bit confused about where we landed not having an exact mapping to a familiar UI element. Here, `stroke("blue", stroke_width=3)` maps to two "symbology mappings" in the UI, and `constant_num(20, target="radius")` maps to one. These two method calls are totally independent of each other, yet they're chained. That leaves room for incorrect interpretation. E.g. why are these method calls chained? What happens if I change the order? This example is particularly strange looking:

```python
    symbology=(
        identity(value="$band-1", target="pixel-red")
        .identity(value="$band-1", target="pixel-green")
        .identity(value="$band-1", target="pixel-blue")
        .identity(value="$band-1", target="pixel-alpha")
    ),
```

Should we be targeting an API that's easier to explain succinctly and with consistent terminology that applies across the GUI and API? @MMesch I need help with the terminology here, I'm just winging it. But I'm something along the lines of:

> a Layer (TBD better name) is a set of independent Mappings with an optional Preprocessor (e.g. heatmap or cluster). A Mapping is a set of 2-4 chained method calls in the following order: an Input (e.g. scalar value like a band or an attribute, or a constant value); an optional Translation (e.g. linear mapping from input values 0,1 to output values 0,10 or a colormap across input values 0,100); an Encoding (e.g. line thickness or fill color); and an optional Filter (e.g. only apply this mapping when value >=10).

This mental model applies to the GUI now, and to apply it to the Python API, I imagine it might look like:

```python
symbology=[  # one layer -- to have multiple layers use list of lists
    constant(20).encoding("radius"),
    constant("blue").encoding("stroke-color"),
    constant(3).encoding("stroke-width"),
],
```

Or a more complex example based on the earthquakes.jGIS example:

```python
symbology=[
	[  # layer 1
        constant(2).encoding("stroke-width"),  # Constructs a `Mapping`
        scalar("mag").linear(in=(3, 9), out=(1, 10)).encoding("size"),  # Constructs a `Mapping`
        constant(RGBA(255, 0, 0, 0.45)).encoding("fill"),  # Constructs a `Mapping`
    ],
    heatmap(radius=20, blur=30, weight="mag", mappings=[  # Constructs `SymbologyLayer` 2
        color_ramp("viridis", attribute="$density").encoding("fill"),  # Constructs a `Mapping`
        constant(0.9).encoding("transparency"),  # Constructs a `Mapping`
    ]),
],
```

This way the chaining has a relational meaning. Since these mappings have a clear structure and order, to me it makes sense to describe that with chained calls, and then we can do more robust validation at each step of the chain. And by using list elements to represent individual mappings, we clearly represent that they are part of a collection, but not inter-related or inter-dependent.

What do you think? Does this feel more intuitive and readable and explainable to you?

Author: I think you have a point that the chaining is not really meaningful right now and could be replaced with a list/tuple.

I am not 100% sure the suggested chaining e.g. `constant(2).encoding("stroke-width")` makes things simpler. If we simply replace the chaining from my PR with simple lists, I find the examples a tiny bit clearer from what you suggested:

```python
symbology=[  # one layer -- to have multiple layers use list of lists
    constant_num(20, target="radius"),
    constant_color("blue", target="stroke-color"),
    constant_num(3, target="stroke-width"),
],
```

or even maybe this if we detect colors automatically:

```python
symbology=[  # one layer -- to have multiple layers use list of lists
    constant(20, target="radius"),
    constant("blue", target="stroke-color"),
    constant(3, target="stroke-width"),
],
```

and for the more complexe example you provided:

```python
symbology=[
	[  # layer 1
        constant(2, target="stroke-width"),  # Constructs a `Mapping`
        scalar("mag", type="linear", input_range=(3, 9), output_range=(1, 10), target="size",  # Constructs a `Mapping`
        constant(RGBA(255, 0, 0, 0.45) target="fill"),  # Constructs a `Mapping`
    ],
    heatmap(radius=20, blur=30, weight="mag", mappings=[  # Constructs `SymbologyLayer` 2
        color_ramp("viridis", attribute="$density", target="fill"),  # Constructs a `Mapping`
        constant(0.9, target="transparency"),  # Constructs a `Mapping`
    ]),
],
```

what do you think?

Author: cc. @MMesch

Reviewer: >what do you think?

I think this is an improvement because it eliminates the ambiguous semantics of chaining in the current implementation, but I also think this is pushing in to parameter overload territory. I think the chaining is a **brilliant** idea if we apply it to mappings, because there's a natural linear relationship between the elements of a mapping. If we use a single constructor per mapping instead of a chain of constructors, especially with the need for some parameters to be dependent or mutually exclusive and the kw parameters can be passed in any order, it will be difficult to document, reason about, read, and validate.

With chaining, we can create discrete objects for each step that enable more robust validation, better error messaging, easier debugging, and better semantics that mirror the GUI. Those semantics will give users the terminology they need to reason about and talk with collaborators about the JupyterGIS environment unambiguously, e.g. "Let's visualize the `temp` **scalar**, **translate** its values from 0 to 8 linearly into the "viridis" colormap space, and **encode** it as pixel color". Someone with those instructions should be able to complete them in the GUI or programmatically, and the latter should be able to be read back in similar language, like `scalar("temp").color_ramp("viridis", interpolate="linear", input_range=(0, 8), output_range=(1, 10)).encode("pixel-color")`. (and for multiple encodings you could `.encode([...])`)

I'm curious what does or does not resonate with you about this proposal? I do think this is a simpler architectural choice because it's easier to describe and constrain what each method in the chain produces and is capable of consuming, and because there will be less messy conditionals required to deal with a large (and growing) number of parameters.

Reviewer: or

```python
(
    scalar("temp")
    .color_ramp(
        "viridis",
        interpolate=LinearInterpolation(input_range=(0, 8), output_range=(1, 10)),
    ).encode("pixel-color")
)
```

Author: Makes sense 👍🏽 The only small worry I have is that halfway through the chain you end up with a half-baked symbology definition (pydantic backed) so we need to make pydantic happy somehow by providing defaults.

Reviewer: What about using different types at each stage to represent incomplete symbology, each containing a reference to their parent and any needed custom validation? Like:

```python
class SymbologyMappingInputScalar(SymbologyMappingInput):
    field_name: str

    @overload
    def translate(self, type=Literal["linear"]) -> SymbologyMappingTranslationLinear:
        pass

    def translate(self, type=Literal["linear", ...]) -> SymbologyMappingTranslation:
        ...

class SymbologyMappingTranslationLinear(SymbologyMappingTranslation):
    input: SymbologyMappingInputScalar  # Can't be categorical, identity, colorramp, or const! Pydantic will validate.
    input_range: tuple[float, float]
    output_range: tuple[float, float]

    def encode(...) -> SymbologyMapping:
        ...

class SymbologyMapping:
    input: SymbologyMappingInput
    translation: SymbologyMappingTranslation
    encoding: SymbologyMappingEncoding
    ...
```

During today's meeting we discussed a need to re-use as much as possible the types generated from schema, and I agree. We don't want a sync problem, that sucks. Though the approach above may have advantages for validation/error handling and inspectability/debugability by users. I'm unsure what the maintenance burden would look like after a full implementation including DRYing, but it might be worth it?

An idea was raised to have some concept of a partial pydantic type (like TypeScript's `Partial<...>` generic). There's a [package](https://pypi.org/project/pydantic-partial/) out there for that, and an [open issue](https://github.com/pydantic/pydantic/issues/1673).

Author: I took your comment into account, please have a look!

doc = GISDocument(zoom=2, latitude=58.118086, longitude=-98.799263)

doc.add_geoparquet_layer(
    path="https://raw.githubusercontent.com/opengeospatial/geoparquet/main/examples/example.parquet",
    name="Parquet example",
    symbology=stroke("blue").fill("lightblue")
)

doc
```

### GeoPackage

```python
from jupytergis import GISDocument, stroke, fill

doc = GISDocument(zoom=4, latitude=45.895322, longitude=2.267552)

doc.add_geopackage_vector_layer(
    path="https://raw.githubusercontent.com/richard-thomas/ol-load-geopackage/master/examples/dist/Natural_Earth_QGIS_layers_and_styles.gpkg",
    name="Geopackage example",
    table_names=["Countries", "Rivers + Lake Centrelines"],
    symbology=stroke("blue").fill("green")
)

doc
```

### Remove layers and inspect document content

```python
layer_id = doc.add_raster_layer(url="https://tile.openstreetmap.org/{z}/{x}/{y}.png")

snapshot = doc.to_py()
print(snapshot.keys())  # layers, sources, layerTree, options, metadata

doc.remove_layer(layer_id)
```

You can also inspect convenience properties:

```python
print(doc.layers)      # dict keyed by layer id
print(doc.layer_tree)  # ordered layer ids / groups
```

> ## explore

Reviewer: Let's pull these API docs into a separate page instead of combining them with the how-to docs we're adding!

# Python API

## GISDocument

The `GISDocument` class is the main entry point for map creation and editing.

### Create a new ephemeral document with map view options

If you don't pass a filename as the first positional argument, the document will not be saved to disk.

```python
from jupytergis import GISDocument

doc = GISDocument(
    latitude=47.0,
    longitude=2.0,
    zoom=5,
    projection="EPSG:3857",
)

doc
```

> If this document doesn't already exist, it will be created on disk.

Reviewer: ```suggestion
```

Woops I think this went in the wrong place. That's probably my fault! This document doesn't have a filename so it will never get created.

### Open an existing document

It is possible to open an existing document and start editing it from the Python notebook.

The Python runtime then acts as a collaborator on the document (in the live collaboration sense), so other people can see the modifications being made to the document.

```python
from jupytergis import GISDocument

doc = GISDocument("mydocument.jGIS")

doc
```

If this document doesn't exist, it will be created on disk.

### Add base layers and vector/raster data

```python
from jupytergis import GISDocument, stroke

doc = GISDocument()

doc.add_raster_layer(
    url="https://tile.openstreetmap.org/{z}/{x}/{y}.png"
)
doc.add_geojson_layer(
    name="Ship Trajectories",
    path="https://openlayers.org/en/latest/examples/data/geojson/ship-trajectories.json",
    symbology=stroke("red")

> )

Reviewer: This example currently doesn't work for me, the ship trajectories layer doesn't display anything! The legend looks OK but not the map.

Author: Same, you're hitting a race condition https://github.com/geojupyter/jupytergis/issues/1481. It has nothing to do with this PR, only that now that we provide more APIs and play with it a bit more we hit it more 😬

doc.add_geotiff_layer(
    name="Sentinel-2 cloudless",
    url="https://s2downloads.eox.at/demo/EOxCloudless/2020/rgbnir/s2cloudless2020-16bits_sinlge-file_z0-4.tif",
    min=2000,
    max=25000,
)

doc
```

### Add GeoJSON from in-memory data

```python
from jupytergis import GISDocument, stroke, constant_num

latitude = 48.853757
longitude = 2.358213

doc = GISDocument(zoom=12, latitude=latitude, longitude=longitude)

geojson = {
    "type": "FeatureCollection",
    "features": [
        {
            "type": "Feature",
            "geometry": {"type": "Point", "coordinates": [longitude, latitude]},
            "properties": {"name": "Paris", "mag": 4.2},
        },
    ],
}

doc.add_geojson_layer(
    name="Paris",
    data=geojson,

>     symbology=stroke("blue", stroke_width=3).constant_num(20, target="radius")

Reviewer: ```
AttributeError: 'Layer' object has no attribute 'constant_num'
```

I think I might be reviewing too early :) This looks like the old style, too.
